### PR TITLE
Vectorize non-persistent Softmax kernels

### PR DESCRIPTION
**Pull request by slayton58**
Add read/write vectorization to non-persistent softmax kernels only. At this point launch logic has minimal changes, and `ILP=vectorization=2` is always used (the code can handle other values, but `ILP=2` has been the most consistent performer).

Dispatch to persistent / non-persistent kernels is unchanged.

### CODE REVIEW

**Comment by ngimel** [comment]
Before and after benchmarks would be useful.

**Comment by ngimel** [comment]
Also, tests with unaligned/oddly sized inputs.

**Comment by slayton58** [comment]
Perf results on my local Quadro RTX6000 (TU102) - shape is the input shape for `x`, offset means that I shift using 
```
x = x[offset[0]:, offset[1]:]
```

Times are for fprop & bprop together, run 50 times with 5 warmup runs: `t = (t_stop - t_start) / (iters - warmup)`

All results in fp16.
```
( shape, offsets) : t (seconds) vectorized : t (seconds) original
===== CORRECTNESS =====
((104, 30064), (2, 1)) : 0.00021066135830349392 : 0.0002200868394639757
((127, 30460), (3, 6)) : 0.00022295316060384114 : 0.0002415445115831163
((109, 30096), (3, 0)) : 0.00015841590033637153 : 0.00016992357042100694
((120, 30443), (4, 0)) : 0.00017747879028320312 : 0.00019176801045735678
((124, 30272), (5, 3)) : 0.0002140204111735026  : 0.00022433069017198352
((102, 30325), (0, 0)) : 0.00016125043233235678 : 0.00016658571031358507
(( 97, 30554), (0, 6)) : 0.00017596350775824654 : 0.00019892056783040365
((109, 30432), (5, 0)) : 0.00016812748379177518 : 0.0001804457770453559
((110, 30448), (3, 3)) : 0.00018289883931477864 : 0.0002128177218967014
((118, 30236), (5, 3)) : 0.0001926528082953559  : 0.00021415816413031683
===== PERF =====
((104, 30064), (0, 0)) : 0.0001567999521891276 : 0.00022451612684461805
((112, 30120), (0, 0)) : 0.00016772482130262588 : 0.00019234551323784723
((124, 30460), (0, 0)) : 0.0001903692881266276 : 0.00020457373725043404
((124, 30388), (0, 0)) : 0.0002003351847330729 : 0.0001956992679172092
((108, 30096), (0, 0)) : 0.00017195807562934027 : 0.00017196867201063368
((124, 30028), (0, 0)) : 0.0001846949259440104 : 0.00019939740498860678
((120, 30440), (0, 0)) : 0.00019349522060818142 : 0.00019965171813964843
((96, 30456), (0, 0)) : 0.00015688472323947483 : 0.00017176734076605903
((112, 30232), (0, 0)) : 0.00017905765109592014 : 0.0001776960160997179
((100, 30324), (0, 0)) : 0.00016157362196180555 : 0.00016380416022406685

**Comment by ngimel** [comment]
Can you please rebase? The CI is just shot.

**Comment by slayton58** [comment]
Ok, rebased from `origin/master`, confirmed building & still running locally

**Comment by slayton58** [comment]
@ngimel ROCM fails look to be related, but I don't have any visibility into the issue: 

```
14:37:42 /var/lib/jenkins/workspace/aten/src/ATen/native/hip/SoftMax.hip:873:10: note: in instantiation of function template specialization 'at::native::(anonymous namespace)::host_softmax_backward<LogSoftMaxBackwardEpilogue, true>' requested here
14:37:42   return host_softmax_backward<LogSoftMaxBackwardEpilogue,true>(grad, output, dim, half_to_float);
14:37:42          ^
14:37:42 /opt/rocm/include/hip/hcc_detail/functional_grid_launch.hpp:173:6: note: candidate function [with Args = <type-parameter-0-1 *, type-parameter-0-3 *, type-parameter-0-3 *, int>, F = void (*)(type-parameter-0-1 *, type-parameter-0-3 *, type-parameter-0-3 *, int)] not viable: no overload of 'cunn_SoftMaxBackward' matching 'void (*)(type-parameter-0-1 *, type-parameter-0-3 *, type-parameter-0-3 *, int)' for 1st argument
14:37:42 void hipLaunchKernelGGL(F kernel, const dim3& numBlocks, const dim3& dimBlocks,
14:37:42      ^
14:37:42 /var/lib/jenkins/workspace/aten/src/ATen/native/hip/SoftMax.hip:819:8: error: no matching function for call to 'hipLaunchKernelGGL'
14:37:42        hipLaunchKernelGGL( cunn_SoftMaxBackward<ILP, scalar_t, accscalar_t, accscalar_t, Epilogue>
```
Nothing has changed in terms of the templates and arguments to `cunn_Softmax{Forward,Backward}` so I'm not sure why this would start failing (if the above error is actually the root of the failure...)

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -616,6 +813,8 @@ Tensor host_softmax_backward(const Tensor &grad_, const Tensor &output_, int64_t
         dispatch_softmax_backward<accscalar_t, scalar_t, accscalar_t, is_log_softmax>(
             gI.data_ptr<scalar_t>(), grad.data_ptr<accscalar_t>(), output.data_ptr<accscalar_t>(), dim_size, dim_size, outer_size);
       } else {
+        const int ILP = sizeof(float4) / std::max(sizeof(scalar_t), sizeof(accscalar_t));
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu:816]: Can you change this to `constexpr` to try if ROCm is fixed?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -606,6 +801,8 @@ Tensor host_softmax_backward(const Tensor &grad_, const Tensor &output_, int64_t
         dispatch_softmax_backward<scalar_t, scalar_t, accscalar_t, is_log_softmax>(
             gI.data_ptr<scalar_t>(), grad.data_ptr<scalar_t>(), output.data_ptr<scalar_t>(), dim_size, dim_size, outer_size);
       } else {
+        const int ILP = sizeof(float4) / sizeof(scalar_t);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/SoftMax.cu:804]: Same here

**Comment by slayton58** [comment]
ROCm is finally happy, XLA test doesn't look related

File: test/test_nn.py [hunk 3]
```diff
@@ -9637,14 +9637,18 @@ def test_embedding_padding_idx(self, device):
     @dtypesIfCUDA(torch.half, torch.float)
     @dtypes(torch.float)
     def test_softmax_backward(self, device, dtype):
-        sizes = [(0, 10), (32, 20), (10, 0)]
+        # Non-even sizes and non-zero shifts test fallback paths in vectorized kernel
+        sizes = [(0, 10), (32, 20), (10, 0), (31, 20), (32, 21), (31, 23)]
```
Comment (ngimel) [inline test/test_nn.py:9641]: Wouldn't these small sizes (<1024) be dispatched to specialized kernel?

**Review by ngimel — APPROVED** [review]

**Comment by ailzhang** [comment]
@slayton58 Would you mind rebasing on top of master? XLA job wasn't able to run due to some merge conflict with master. Thanks a lot!

File: test/test_nn.py (hunk 3 above)
Comment (slayton58) [reply test/test_nn.py]: Added new tests

**Comment by slayton58** [comment]
@ailzhang Rebased and pushed :)

**Comment by ailzhang** [comment]
Thanks a lot @slayton58 ! Btw I'm curious why your PR hit merge conflict in the jobs while github doesn't complain? I suspect what we do here in https://github.com/pytorch/pytorch/blob/master/.circleci/config.yml#L128-L139 has bugs but not sure. Please let me know if you have an idea what could be the root cause for it. Thanks!

**Comment by ppwwyyxx** [comment]
This PR does not compute correct outputs:
```
╭─ 
╰─$ a = torch.rand(32, 65537).cuda()

╭─ 
╰─$ a.log_softmax(dim=1)
Out[13]: 
tensor([[-10.7702, -11.0043, -11.4230,  ..., -10.6436, -11.5849, -11.5748],
        [-11.5994, -11.0413, -11.1629,  ..., -11.4148, -10.9850,   0.6423],
        [-10.6430, -10.9107, -11.2595,  ..., -10.8940,   0.3540,   0.5523],
        ...,
        [-11.4674, -11.5610, -10.7308,  ..., -10.6490, -11.4639,   0.7431],
        [-10.7557, -11.4076, -11.4337,  ..., -10.8813,   0.0935,   0.7081],
        [-10.9402, -11.2714, -11.4213,  ...,   0.5622,   0.9081,   0.4131]],
       device='cuda:0')

╭─ 
╰─$ torch.log(torch.exp(a) / torch.exp(a).sum(dim=1, keepdim=True))
Out[14]: 
tensor([[-10.7702, -11.0043, -11.4230,  ..., -10.6436, -11.5849, -11.5748],
        [-11.5994, -11.0413, -11.1629,  ..., -11.4148, -10.9850, -11.0052],
        [-10.6430, -10.9107, -11.2595,  ..., -10.8940, -10.9034, -10.8701],
        ...,
        [-11.4674, -11.5610, -10.7308,  ..., -10.6490, -11.4639, -11.3691],
        [-10.7557, -11.4076, -11.4337,  ..., -10.8813, -11.3089, -10.6395],
        [-10.9402, -11.2714, -11.4213,  ..., -11.3528, -11.0477, -10.8662]],
       device='cuda:0')
```

**Comment by ngimel** [comment]
Sorry, I'll unland now.

**Comment by soumith** [comment]
i'm reverting in D21589011, should have an exported PR in a moment.

**Comment by zasdfgbnm** [comment]
Does this happen only for `65537` and larger?

**Comment by ngimel** [comment]
Nope, just any odd-sized tensor (tried 35537)

**Comment by slayton58** [comment]
Sorry about this -- fixed on my end (1 line each in two kernels, thanks @ngimel for the pointer)
```
# python softmax_test.py 
tensor([[-13.2594, -10.6079, -10.2999,  ..., -12.4991, -11.2978, -15.0369],
        [ -9.9151, -11.4201, -11.2289,  ..., -12.0209, -11.8791, -12.1396],
        [-14.1754, -11.6598, -12.3467,  ..., -11.0960, -12.9246, -12.6462],
        ...,
        [-11.7451, -13.7147, -12.5571,  ..., -11.3787, -12.0840, -12.1193],
        [-10.8610, -11.7080,  -9.9343,  ..., -12.7462, -14.1307, -12.8460],
        [-12.6018, -11.6838, -12.2617,  ..., -11.5191, -11.1250, -12.6795]],
       device='cuda:0', grad_fn=<LogSoftmaxBackward>)
tensor([[-13.2594, -10.6079, -10.2999,  ..., -12.4991, -11.2978, -15.0369],
        [ -9.9151, -11.4201, -11.2289,  ..., -12.0209, -11.8791, -12.1396],
        [-14.1754, -11.6598, -12.3467,  ..., -11.0960, -12.9246, -12.6462],
        ...,
        [-11.7451, -13.7147, -12.5571,  ..., -11.3787, -12.0840, -12.1193],
        [-10.8610, -11.7080,  -9.9343,  ..., -12.7462, -14.1307, -12.8460],
        [-12.6018, -11.6838, -12.2617,  ..., -11.5191, -11.1250, -12.6795]],
       device='cuda:0', grad_fn=<LogBackward>)
=======
tensor([[ 0.9216, -0.4311, -0.0936,  ..., -1.0507, -0.8658,  0.7814],
        [-0.8343,  1.6185,  0.4378,  ..., -2.1977, -1.5487, -0.3499],
        [ 0.1788, -0.8722, -0.7476,  ...,  0.3836, -0.4154,  0.1673],
        ...,
        [ 0.9146,  0.4633,  1.3884,  ..., -1.8522, -0.1852,  0.3311],
        [-0.8222,  0.3677, -0.0766,  ..., -1.0111,  0.4705,  0.2457],
        [-1.4741, -0.9013,  0.9225,  ..., -0.9396, -0.7337, -0.2698]],
       device='cuda:0')
tensor([[ 0.9216, -0.4311, -0.0936,  ..., -1.0507, -0.8658,  0.7814],
        [-0.8343,  1.6185,  0.4378,  ..., -2.1977, -1.5487, -0.3499],
        [ 0.1788, -0.8722, -0.7476,  ...,  0.3836, -0.4154,  0.1673],
        ...,
        [ 0.9146,  0.4633,  1.3884,  ..., -1.8522, -0.1852,  0.3311],
        [-0.8222,  0.3677, -0.0766,  ..., -1.0111,  0.4705,  0.2457],
        [-1.4741, -0.9013,  0.9225,  ..., -0.9396, -0.7337, -0.2698]],
       device='cuda:0')
```

test_softmax.py is effectively the code from https://github.com/pytorch/pytorch/pull/36485#issuecomment-629003067 with additional grad computation.

**Comment by slayton58** [comment]
Test code I used: https://gist.github.com/slayton58/14f3fd176f32064731638baf789504cd
